syntax = "proto3";

package protos;



import "google/protobuf/empty.proto";
import "google/protobuf/timestamp.proto";
import "custom_types.proto";

service CustomerService {
  rpc GetAll (GetAllRequest) returns (GetAllCustomersResponse);
  rpc GetById (GetCustomerByIdRequest) returns (CustomerModel);
  rpc GetBySubstring (GetBySubstringRequest) returns (GetAllCustomersResponse);
  rpc GetByDateRange (GetByDateRangeRequest) returns (GetAllCustomersResponse);
}

message CustomerModel {
  int32 customer_id = 1;
  string first_name = 2;
  string last_name = 3;
  string email = 4;
  optional string phone = 5;
  optional DateOnly birth_date = 6;
  optional google.protobuf.Timestamp created_at = 7;
}

message GetBySubstringRequest{
  string sub_str = 1;
  optional int32 page_n = 2;
  optional int32 page_size = 3;
  string order = 4;
  optional bool desc = 5;
}



message GetAllCustomersResponse {
  repeated CustomerModel customers = 1;
}

message GetCustomerByIdRequest {
  int32 customer_id = 1;
}
